contracts/style.md

```diff
@@ -0,0 +1,215 @@
+
+- Control flow modifiers: Prefix the modifier name with `if` in the case that a modifier only enables or disables the subsequent code in the modified method, but does not revert.
+- Reverting modifiers: Prefix the modifier name with `validate` in the case that a modifier reverts if a condition is not met.
+
+### Return Values
+
+- If an address is cast as a contract type, return the type, do not cast back to the address type. This prevents the consumer of the method signature from having to cast again, but presents an equivalent API for off-chain APIs. Additionally it is a more declarative API, providing more context if we return a type.
+
+## Events
+
+- Events should only be triggered on state changes. If the value is set but not changed, we prefer avoiding a log emission indicating a change. (e.g. Either do not emit a log, or name the event `ConfigSet` instead of `ConfigUpdated`.)
+
+### Naming
+
+- When possible event names should correspond to the method they are in or the action that is being taken. Events preferably follow the format <subject><actionPerformed>, where the action performed is the past tense of the imperative verb in the method name.  e.g. calling `setConfig` should emit an event called `ConfigSet`, not `ConfigUpdated` in a method named `setConfig`.
+
+## Errors
+
+### Use Custom Errors
+
+Whenever possible (Solidity v0.8+) use [custom errors](https://blog.soliditylang.org/2021/04/21/custom-errors/) instead of emitting strings. This saves contract code size and simultaneously provides more informative error messages.
+
+### Expose Errors
+
+It is common to call a contract and then check the call succeeded:
+
+```javascript
+(bool success, ) = to.call(data);
+require(success, "Contract call failed");
+```
+
+While this may look descriptive it swallows the error. Instead bubble up the error:
+
+```javascript
+error YourError(bytes response);
+
+(bool success, bytes memory response) = to.call(data);
+if (!success) revert YourError(response);
+```
+
+This will cost slightly more gas to copy the response into memory, but will ultimately make contract usage more understandable and easier to debug. Whether it is worth the extra gas is a judgement call you’ll have to make based on your needs.
+
+The original error will not be human readable in an off-chain explorer because it is RLP hex encoded but is easily decoded with standard Solidity ABI decoding tools, or a hex to UTF-8 converter and some basic ABI knowledge.
+
+## Interfaces
+
+### Scope
+
+- Interfaces should be as concise as reasonably possible. Break it up into smaller composable interfaces when that is sensible.
+
+### Naming
+
+- Interfaces should be named `IFoo` instead of `FooInterface`. This follows the patterns of popular [libraries like OpenZeppelin’s](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/IERC20.sol#L9).
```

Comment:
Given that we're likely not applying this rule until Sol v0.9 contracts are created, should we indicate that here or hold off merging this until then?